controllers/twophase/state_machine.go

```diff
@@ -0,0 +1,173 @@
+	currentPhase := m.Chaos.GetStatus().Experiment.Phase
+	m.Log.Info("change phase", "current phase", currentPhase, "target phase", targetPhase)
+
+	// covers 6 conditions
+	if currentPhase == targetPhase {
+		return updated, nil
+	}
+
+	// covers 5 conditions
+	if currentPhase == v1alpha1.ExperimentPhaseFinished {
+		return updated, unexpectedInto(currentPhase, targetPhase)
+	}
+
+	// In this branch `targetPhase != v1alpha1.ExperimentPhaseRunning` is always true
+	// move from Running phase to any phase should recover the experiment.
+	// covers 5 conditions
+	if currentPhase == v1alpha1.ExperimentPhaseRunning {
+		updated = true
+
+		m.Log.Info("recovering", "current phase", currentPhase, "target phase", targetPhase)
+		err := m.Recover(ctx, m.Req, m.Chaos)
+		if err != nil {
+			status.FailedMessage = err.Error()
+
+			m.Log.Error(err, "fail to recover")
+			return updated, err
+		}
+		status.Experiment.Phase = targetPhase
+
+		now := time.Now()
+		status.Experiment.EndTime = &metav1.Time{
+			Time: now,
+		}
+		status.Experiment.Duration = now.Sub(status.Experiment.StartTime.Time).String()
+
+		// Pause from a running phase should set the nextStart to now
+		// so that the Reconciler will start another cycle of running right
+		// after resume
+		if targetPhase == v1alpha1.ExperimentPhasePaused {
+			m.Chaos.SetNextStart(now)
+		}
+	}
+
+	// currentPhase could be "" or Waiting or Paused or Failed
+	// covers 4 * 5 = 20 conditions
+	if currentPhase != v1alpha1.ExperimentPhaseRunning {
+		if targetPhase == v1alpha1.ExperimentPhaseRunning {
+			updated = true
+
+			m.Log.Info("applying", "current phase", currentPhase, "target phase", targetPhase)
+			err := m.Apply(ctx, m.Req, m.Chaos)
+			if err != nil {
+				m.Log.Error(err, "fail to apply")
+
+				status.Experiment.Phase = v1alpha1.ExperimentPhaseFailed
+				status.FailedMessage = err.Error()
+
+				return updated, err
+			}
+			status.Experiment.Phase = targetPhase
+
+			startTime := time.Now()
+			nextStart, err := utils.NextTime(*m.Chaos.GetScheduler(), startTime)
+			if err != nil {
+				m.Log.Error(err, "failed to get the next start time")
+				return updated, err
+			}
+			nextRecover := startTime.Add(*duration)
+			// if the duration is too long, `nextRecover` could be after `nextStart`
+			// we can jump over a start to make sure `nextRecover` is before `nextStart`
+			for nextRecover.After(*nextStart) {
+				nextStart, err = utils.NextTime(*m.Chaos.GetScheduler(), *nextStart)
+				if err != nil {
+					m.Log.Error(err, "failed to get the next start time")
+					return updated, err
+				}
+			}
+
+			m.Chaos.SetNextStart(*nextStart)
+			m.Chaos.SetNextRecover(nextRecover)
+
+			status.Experiment.StartTime = &metav1.Time{Time: startTime}
+			status.Experiment.Duration = duration.String()
+		} else {
+			updated = true
+
+			status.Experiment.Phase = targetPhase
+		}
+	}
+
+	if status.FailedMessage != emptyString {
+		updated = true
+	}
+	status.FailedMessage = emptyString
```

Comment:
why need to do this?

Reply:
Remove the `failedMessage` from the former execution.